index.js

```diff
@@ -548,20 +605,72 @@ class KeyringController extends EventEmitter {
    * @param {string} password - The keyring controller password.
    * @returns {Promise<Array<Keyring>>} The keyrings.
    */
-  async unlockKeyrings(password) {
+  async unlockKeyrings(password, encryptedKey) {
     const encryptedVault = this.store.getState().vault;
     if (!encryptedVault) {
       throw new Error('Cannot unlock without a previous vault.');
     }
 
     await this.clearKeyrings();
-    const vault = await this.encryptor.decrypt(password, encryptedVault);
-    this.password = password;
+
+    // MV3: If the separator string is in the vault string, the user has already migrated
+    // from the previous password-only model
+    let vault = null;
+    if (encryptedVault?.includes?.(VAULT_SEPARATOR)) {
+      const { salt, vault: vaultOnly } = this.parseVault(encryptedVault);
+
+      log('[unlockKeyrings] salt is: ', salt, '; encryptedVault is:', encryptedVault);
+
+      if (password !== undefined) {
+        this.encryptedKey = await this._generateEncryptedKey(password, salt);
+
+        log('[unlockKeyrings] *new* encryptedKey is: ', this.encryptedKey);
+
+      } else if (encryptedKey !== undefined) {
+        log('[unlockKeyrings] *old* encryptedKey is: ', this.encryptedKey);
+        this.encryptedKey = encryptedKey;
+      } else {
+        throw new Error(
+          'No way to decrypt a salted vault without a password or encrypted key',
+        );
+      }
+
+      vault = await this.encryptor.decrypt(this.encryptedKey, vaultOnly);
+    } else {
+      vault = await this.encryptor.decrypt(password, encryptedVault);
+    }
+
     await Promise.all(vault.map(this._restoreKeyring.bind(this)));
+
     await this._updateMemStoreKeyrings();
+
+    // MV3: If we're provided a password, we should persist keyrings
+    // so that we can either (1) migrate or (2) create a new salt
```

Comment:
I think the below if block should probably go in `submitPassword`, after `unlockKeyrings` resolves